lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -837,6 +837,65 @@ bool firrtl::areTypesWeaklyEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
          destFlip == srcFlip;
 }
 
+/// Returns whether the srcType can be const-casted to the destType.
+bool firrtl::areTypesConstCastable(FIRRTLType destFType, FIRRTLType srcFType,
+                                   bool srcOuterTypeIsConst) {
+  // Identical types are always castable
+  if (destFType == srcFType)
+    return true;
+
+  auto destType = destFType.dyn_cast<FIRRTLBaseType>();
+  auto srcType = srcFType.dyn_cast<FIRRTLBaseType>();
+
+  // For non-base types, only castable if identical.
+  if (!destType || !srcType)
+    return false;
+
+  // Cannot cast non-'const' src to 'const' dest
+  if (destType.isConst() && !srcType.isConst() && !srcOuterTypeIsConst)
```

Comment:
Consider just stashing `srcType.isConst() || srcOuterTypeIsConst` into a variable, and using `!srcIsEffectivelyConst` (or whatever) here (and in the recursion below)?

Reply:
Done